extmod/modopenamp.c

```diff
@@ -411,6 +422,6 @@ const mp_obj_module_t openamp_module = {
 };
 
 MP_REGISTER_ROOT_POINTER(struct _virtio_dev_obj_t *virtio_device);
-MP_REGISTER_MODULE(MP_QSTR_openamp, openamp_module);
+MP_REGISTER_EXTENSIBLE_MODULE(MP_QSTR_openamp, openamp_module);
```

Comment:
Why do you want to make this module extensible?  In general, this feature is only used to support CPython modules that we don't fully support in C, so they can be extended in Python.

For the `openamp` module we define it ourselves, so there's nothing missing that needs adding.

Reply:
I extend in Python to add utils and classes that can be accessed with `openamp.x`. For example I have an `EndPointIO` class which is a subclass of `IOBase` that you can use for `dupterm`.

Reply:
Hm I guess my `EndpointIO` class would not be needed if https://github.com/micropython/micropython/pull/14181 is merged but I still have other extensions.

Reply:
> Note that there's an openamp.Endpoint in the C-module, so openamp.EndpointIO belongs there too but as an extension.

I don't think it's a good idea to add things to this module outside this repo.  We should fully define the `openamp` module here, and have the docs here.  Any extensions that work with `openamp` but are outside this repo (which would be third-party extensions, relative to this repo) should be placed in a separate module, eg `openamp_ext`, with separate docs.  Otherwise it will get confusing what functions/classes are available in this module.

Eventually we want to get rid of this extensible module stuff (it's a backwards compatible way to support the old u-module naming scheme), so I don't want to add more modules that use this.

Reply:
> Eventually we want to get rid of this extensible module stuff (it's a backwards compatible way to support the old u-module naming scheme), so I don't want to add more modules that use this.

Whoa that will be a very breaking change. We heavily depend on extending modules for adding missing functionality or for portability. For example, we extend machine to add an LED that's used throughout or examples. We'll also extend it to mask missing machine.PWM for `stm32` port. Can this feature be kept for such uses cases ?

Reply:
Did I understand correctly that we'll no longer be able to extend C-Modules from Python ? Note that we've switched to using the extensible macro a while ago, so we have accumulated a bit of an ecosystem around it, extensions, example scripts and docs. I'm mentioning this because I really think decisions to remove things like this should consider the side-effects to users of the C-API. It's been hard to keep up with the C-API breaking changes, but it's all for the greater good. However, I don't understand why a feature like this should be removed. We maintain a fork anyway, we can adapt, but I've been working really hard to align with the upstream, in the hope that one day we can ditch the fork.

Reply:
> Eventually we want to get rid of this extensible module stuff (it's a backwards compatible way to support the old u-module naming scheme), so I don't want to add more modules that use this.

Sorry, I didn't really express myself clearly in this comment.

What I intended to say is that we want to eventually drop the u-module naming backwards compatibility feature, which is tied in to the extensible module macro.  But you're right that we can't get rid of the ability to extend C modules from Python.

At the same time, I think it's important to limit the number of extensible modules because  extensible ones are slower to import than non-extensible ones, due to needing to search the filesystem first.  So their use should be limited and only when necessary.

You can see that the extensible modules are strictly defined and documented here: https://docs.micropython.org/en/latest/library/index.html#extending-built-in-libraries-from-python .  We really do want to limit their use.

See also the comment in `extmod/modbluetooth.c`:
```c
// This module should not be extensible (as it is not a CPython standard
// library nor is it necessary to override from the filesystem), however it
// has previously been known as `ubluetooth`, so by making it extensible the
// `ubluetooth` alias will continue to work.
MP_REGISTER_EXTENSIBLE_MODULE(MP_QSTR_bluetooth, mp_module_bluetooth);
```

For the case here of `openamp`, I really think it should not be extensible.  If there's important additional C functionality then it should be added in this repo, with docs, so it's available on all ports.  Or if that functionality is best implemented in Python, and it makes more sense to have it in `openamp` rather than a separate module, then it should go in `micropython-lib` and then, yes, the module would be extensible.

IMO it's best to limit the scope of the low-level `openamp` module to just the necessary C functionality, and then have a different higher-level module written in Python that uses `openamp` to provide more functionality.  For example that's done with `_thread` and `threading`, `bluetooth` and `aioble`, `machine.USBDevice` and `usb-device`, to name a few.

Reply:
> Note that we've switched to using the extensible macro a while ago, so we have accumulated a bit of an ecosystem around it, extensions, example scripts and docs

That's interesting... this is not really what the feature was intended for.  As I mention above, I think modules should have well defined scope and built-in C modules left to provide basic functionality, while other Python modules build upon the low-level modules and provide more features.

Reply:
> I think this is all subjective

This is a tricky topic for sure, but I suggest the measurable aspect of what Damien is talking about is consistency. The extension mechanism is used to ensure consistency between MicroPython and CPython (i.e. extending os with `os.path`), or between different MP ports (i.e. extending `machine.Pin` to the unix port). In both cases there are docs for those modules already, and the extensions mechanism keeps the implementation (and therefore user experience) as consistent with those docs as possible.

> However, the rest of the extensions, which BTW are Not yet ready for sharing, may or may not be welcome by the maintainers. 

Either way, I think the preferred path for MicroPython and user experience would be to have a separate module here:

* If the additional Python-based functionality gets merged to micropython-lib, then this sounds like it fits well with the common pattern Damien already mentioned of bluetooth/aioble, machine.USBDevice/usb-device, etc.
* If the additional Python functionality is not merged to micropython-lib, then using the same module name would cause the MicroPython docs for the `openamp` module to be inconsistent with the docs for the external extended `openamp` module. Using a different name removes the ambiguity and improves the user experience.

Reply:
> If the additional Python functionality is not merged to micropython-lib, then using the same module name would cause the MicroPython docs for the openamp module to be inconsistent with the docs

But what if we don't want to push our extensions to `micropython-lib` even if they are in line with the upstream policy and will be merged ? Maybe we'd want to maintain them privately, or push them somewhere else, like `arduino-lib` and then we could also maintain our own docs for the module.
The reason for sending work like this upstream, in addition to contributing something useful, is to Not have fork and maintain it ourselves, but if we still have to fork and change the code that we ourselves put upstream just to be able to extend it, then what's the point of pushing it upstream in the first place.

Reply:
Note that you can drop the extensible commit (it's in its own commit) if you don't want it, if this is ready to be merged.

Reply:
> then we could also maintain our own docs for the module.

It's of course your choice if you want to do that. I'd note that if there are two modules called `openamp` with different functions in them, and different docs for that module in two places, then this can easily lead to a bad user experience. i.e. Finding the docs for `openamp` (even `openamp` docs for the matching MP version) won't necessarily find the correct docs, because there will be two different implementations with the same name.

Sometimes that's unavoidable because of external constraints (and the history of MP has plenty of confusing incidents arising from those unavoidable name clashes). In this case it seems like it is avoidable, by giving different modules different names. Maybe I don't understand the use case here, though.

> Note that you can drop the extensible commit (it's in its own commit) if you don't want it, if this is ready to be merged.

:thumbsup: